Fix Native signature for optional Tensor arguments

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50767 Fix Native signature for optional Tensor arguments**

The native signature for optional tensor arguments wrongly produced "Tensor" instead of "optional<Tensor>". We didn't notice this because all internal ops currently use hacky_wrapper, and for hacky_wrapper, "Tensor" is correct.

This PR fixes that and ports one op (batch_norm) to not use hacky_wrapper anymore as a proof of fix.

Differential Revision: [D25960941](https://our.internmc.facebook.com/intern/diff/D25960941/)

smessmer committed:
Fix Native signature for optional Tensor arguments

The native signature for optional tensor arguments wrongly produced "Tensor" instead of "optional<Tensor>". We didn't notice this because all internal ops currently use hacky_wrapper, and for hacky_wrapper, "Tensor" is correct.

This PR fixes that and ports one op (batch_norm) to not use hacky_wrapper anymore as a proof of fix.

Differential Revision: [D25960941](https://our.internmc.facebook.com/intern/diff/D25960941/)

[ghstack-poisoned]

bhosmer commented on tools/codegen/api/native.py:
```diff
@@ -29,10 +29,16 @@ def name(func: FunctionSchema) -> str:
 
 def argumenttype_type(t: Type, *, mutable: bool, binds: ArgName) -> CType:
     if str(t) == 'Tensor?':
```
nit: I'd factor out `tensor_type = BaseCType('Tensor', binds)` for clarity, if you're in here again before landing

smessmer committed:
Update on "Fix Native signature for optional Tensor arguments"


The native signature for optional tensor arguments wrongly produced "Tensor" instead of "optional<Tensor>". We didn't notice this because all internal ops currently use hacky_wrapper, and for hacky_wrapper, "Tensor" is correct.

This PR fixes that and ports one op (batch_norm) to not use hacky_wrapper anymore as a proof of fix.

Differential Revision: [D25960941](https://our.internmc.facebook.com/intern/diff/D25960941/)

[ghstack-poisoned]
